File: Python/Komputer.py
from Cpu import Cpu
from Motherboard import Motherboard
from Ram import Ram
from SSD import SSD
from HDD import HDD
from GPU import GPU
from PowerSupply import PowerSupply
from Casing import Casing

# Kelas Komputer yang menggunakan komposisi dari banyak komponen
class Komputer:
    def __init__(self, nama="", cpu=None, motherboard=None, ram_list=None,
                 gpu=None, power_supply=None, casing=None):
        self.nama = nama
        self.cpu = cpu if cpu is not None else Cpu()
        self.motherboard = motherboard if motherboard is not None else Motherboard()
        self.ram_list = ram_list if ram_list is not None else []
        self.ssd_list = []
        self.hdd_list = []
        self.gpu = gpu if gpu is not None else GPU()
        self.power_supply = power_supply if power_supply is not None else PowerSupply()
        self.casing = casing if casing is not None else Casing()
        self.total_harga = 0
        self.hitung_total_harga()
        
    def set_nama(self, nama):
        self.nama = nama
        
    def set_cpu(self, cpu):
        self.cpu = cpu
        self.hitung_total_harga()
        
    def set_motherboard(self, motherboard):
        self.motherboard = motherboard
        self.hitung_total_harga()
        
    def set_ram_list(self, ram_list):
        self.ram_list = ram_list
        self.hitung_total_harga()
        
    def add_ram(self, ram):
        self.ram_list.append(ram)
        self.hitung_total_harga()
        
    def set_ssd_list(self, ssd_list):
        self.ssd_list = ssd_list
        self.hitung_total_harga()
        
    def add_ssd(self, ssd):
        self.ssd_list.append(ssd)
        self.hitung_total_harga()
        
    def set_hdd_list(self, hdd_list):
        self.hdd_list = hdd_list
        self.hitung_total_harga()
        
    def add_hdd(self, hdd):
        self.hdd_list.append(hdd)
        self.hitung_total_harga()
        
    def set_gpu(self, gpu):
        self.gpu = gpu
        self.hitung_total_harga()
        
    def set_power_supply(self, power_supply):
        self.power_supply = power_supply
        self.hitung_total_harga()
        
    def set_casing(self, casing):
        self.casing = casing
        self.hitung_total_harga()
        
    def get_nama(self):
        return self.nama
        
    def get_cpu(self):
        return self.cpu
        
    def get_motherboard(self):
        return self.motherboard
        
    def get_ram_list(self):
        return self.ram_list
        
    def get_ssd_list(self):
        return self.ssd_list
        
    def get_hdd_list(self):
        return self.hdd_list
        
    def get_gpu(self):
        return self.gpu
        
    def get_power_supply(self):
        return self.power_supply
        
    def get_casing(self):
        return self.casing
        
    def get_total_harga(self):
        return self.total_harga
        
    def hitung_total_harga(self):
        total = self.cpu.get_harga() + self.motherboard.get_harga() + self.gpu.get_harga() + \
                self.power_supply.get_harga() + self.casing.get_harga()
                
        for ram in self.ram_list:
            total += ram.get_harga()
            
        for ssd in self.ssd_list:
            total += ssd.get_harga()
            
        for hdd in self.hdd_list:
            total += hdd.get_harga()
            
        self.total_harga = total
        
    def display_info(self):
        print("==========================================")
        print("           INFORMASI KOMPUTER")
        print("==========================================")
        print(f"Nama Komputer  : {self.nama}")
        print(f"Total Harga    : Rp {self.total_harga}")
        print("==========================================")
        
        print("\n>> CPU:")
        self.cpu.display_info()
        
        print("\n>> Motherboard:")
        self.motherboard.display_info()
        
        print(f"\n>> RAM ({len(self.ram_list)} modul):")
        for i, ram in enumerate(self.ram_list, 1):
            print(f"  RAM #{i}:")
            ram.display_info()
            print()
            
        if self.ssd_list:
            print(f"\n>> SSD ({len(self.ssd_list)} buah):")
            for i, ssd in enumerate(self.ssd_list, 1):
                print(f"  SSD #{i}:")
                ssd.display_info()
                print()
                
        if self.hdd_list:
            print(f"\n>> HDD ({len(self.hdd_list)} buah):")
            for i, hdd in enumerate(self.hdd_list, 1):
                print(f"  HDD #{i}:")
                hdd.display_info()
                print()
                
        print("\n>> GPU:")
        self.gpu.display_info()
        
        print("\n>> Power Supply:")
        self.power_supply.display_info()
        
        print("\n>> Casing:")
        self.casing.display_info()
        
        print("\n==========================================")
        print(f"Total Harga: Rp {self.total_harga}")
        print("==========================================")

File: Java/Komputer.java
import java.util.ArrayList;
import java.util.List;

// Kelas Komputer yang menggunakan komposisi dari banyak komponen
public class Komputer {
    private String nama;
    private Cpu cpu;
    private Motherboard motherboard;
    private List<Ram> ramList;
    private List<SSD> ssdList;
    private List<HDD> hddList;
    private GPU gpu;
    private PowerSupply powerSupply;
    private Casing casing;
    private int totalHarga;

    public Komputer() {
        this.nama = "";
        this.ramList = new ArrayList<>();
        this.ssdList = new ArrayList<>();
        this.hddList = new ArrayList<>();
        this.totalHarga = 0;
    }

    public Komputer(String nama, Cpu cpu, Motherboard motherboard, List<Ram> ramList,
                   GPU gpu, PowerSupply powerSupply, Casing casing) {
        this.nama = nama;
        this.cpu = cpu;
        this.motherboard = motherboard;
        this.ramList = ramList;
        this.ssdList = new ArrayList<>();
        this.hddList = new ArrayList<>();
        this.gpu = gpu;
        this.powerSupply = powerSupply;
        this.casing = casing;
        hitungTotalHarga();
    }

    public void setNama(String nama) {
        this.nama = nama;
    }

    public void setCpu(Cpu cpu) {
        this.cpu = cpu;
        hitungTotalHarga();
    }

    public void setMotherboard(Motherboard motherboard) {
        this.motherboard = motherboard;
        hitungTotalHarga();
    }

    public void setRamList(List<Ram> ramList) {
        this.ramList = ramList;
        hitungTotalHarga();
    }

    public void addRam(Ram ram) {
        this.ramList.add(ram);
        hitungTotalHarga();
    }

    public void setSsdList(List<SSD> ssdList) {
        this.ssdList = ssdList;
        hitungTotalHarga();
    }

    public void addSsd(SSD ssd) {
        this.ssdList.add(ssd);
        hitungTotalHarga();
    }

    public void setHddList(List<HDD> hddList) {
        this.hddList = hddList;
        hitungTotalHarga();
    }

    public void addHdd(HDD hdd) {
        this.hddList.add(hdd);
        hitungTotalHarga();
    }

    public void setGpu(GPU gpu) {
        this.gpu = gpu;
        hitungTotalHarga();
    }

    public void setPowerSupply(PowerSupply powerSupply) {
        this.powerSupply = powerSupply;
        hitungTotalHarga();
    }

    public void setCasing(Casing casing) {
        this.casing = casing;
        hitungTotalHarga();
    }

    public String getNama() {
        return this.nama;
    }

    public Cpu getCpu() {
        return this.cpu;
    }

    public Motherboard getMotherboard() {
        return this.motherboard;
    }

    public List<Ram> getRamList() {
        return this.ramList;
    }

    public List<SSD> getSsdList() {
        return this.ssdList;
    }

    public List<HDD> getHddList() {
        return this.hddList;
    }

    public GPU getGpu() {
        return this.gpu;
    }

    public PowerSupply getPowerSupply() {
        return this.powerSupply;
    }

    public Casing getCasing() {
        return this.casing;
    }

    public int getTotalHarga() {
        return this.totalHarga;
    }

    public void hitungTotalHarga() {
        int total = cpu.getHarga() + motherboard.getHarga() + gpu.getHarga() +
                   powerSupply.getHarga() + casing.getHarga();

        for (Ram ram : ramList) {
            total += ram.getHarga();
        }

        for (SSD ssd : ssdList) {
            total += ssd.getHarga();
        }

        for (HDD hdd : hddList) {
            total += hdd.getHarga();
        }

        this.totalHarga = total;
    }

    public void displayInfo() {
        System.out.println("==========================================");
        System.out.println("           INFORMASI KOMPUTER");
        System.out.println("==========================================");
        System.out.println("Nama Komputer  : " + this.nama);
        System.out.println("Total Harga    : Rp " + this.totalHarga);
        System.out.println("==========================================");

        System.out.println("\n>> CPU:");
        this.cpu.displayInfo();

        System.out.println("\n>> Motherboard:");
        this.motherboard.displayInfo();

        System.out.println("\n>> RAM (" + this.ramList.size() + " modul):");
        for (int i = 0; i < this.ramList.size(); i++) {
            System.out.println("  RAM #" + (i + 1) + ":");
            this.ramList.get(i).displayInfo();
            System.out.println();
        }

        if (!this.ssdList.isEmpty()) {
            System.out.println("\n>> SSD (" + this.ssdList.size() + " buah):");
            for (int i = 0; i < this.ssdList.size(); i++) {
                System.out.println("  SSD #" + (i + 1) + ":");
                this.ssdList.get(i).displayInfo();
                System.out.println();
            }
        }

        if (!this.hddList.isEmpty()) {
            System.out.println("\n>> HDD (" + this.hddList.size() + " buah):");
            for (int i = 0; i < this.hddList.size(); i++) {
                System.out.println("  HDD #" + (i + 1) + ":");
                this.hddList.get(i).displayInfo();
                System.out.println();
            }
        }

        System.out.println("\n>> GPU:");
        this.gpu.displayInfo();

        System.out.println("\n>> Power Supply:");
        this.powerSupply.displayInfo();

        System.out.println("\n>> Casing:");
        this.casing.displayInfo();

        System.out.println("\n==========================================");
        System.out.println("Total Harga: Rp " + this.totalHarga);
        System.out.println("==========================================");
    }
}
